swarm/src/behaviour.rs

```diff
@@ -721,3 +838,214 @@ impl Default for CloseConnection {
+    /// [`InEvent::ConnectionEstablished`] has previously been received with this `PeerId`.
+    ConnectionHandler {
+        peer_id: PeerId,
+        connection: ConnectionId,
+        event: <<Handler as IntoConnectionHandler>::Handler as ConnectionHandler>::OutEvent,
+    },
+    /// Indicates to the behaviour that the dial to a known or unknown node failed.
+    DialFailure {
+        peer_id: Option<PeerId>,
+        handler: Handler,
+        error: &'a DialError,
+    },
+    /// Indicates to the behaviour that an error happened on an incoming connection during its
+    /// initial handshake.
+    ///
+    /// This can include, for example, an error during the handshake of the encryption layer, or the
+    /// connection unexpectedly closed.
+    ListenFailure {
+        local_addr: &'a Multiaddr,
+        send_back_addr: &'a Multiaddr,
+        handler: Handler,
+    },
+    /// Indicates to the behaviour that a new listener was created.
+    NewListener { listener_id: ListenerId },
+    /// Indicates to the behaviour that we have started listening on a new multiaddr.
+    NewListenAddr {
+        listener_id: ListenerId,
+        addr: &'a Multiaddr,
+    },
+    /// Indicates to the behaviour that a multiaddr we were listening on has expired,
+    /// which means that we are no longer listening on it.
+    ExpiredListenAddr {
+        listener_id: ListenerId,
+        addr: &'a Multiaddr,
+    },
+    /// A listener experienced an error.
+    ListenerError {
+        listener_id: ListenerId,
+        err: &'a (dyn std::error::Error + 'static),
+    },
+    /// A listener closed.
+    ListenerClosed {
+        listener_id: ListenerId,
+        reason: Result<(), &'a std::io::Error>,
+    },
+    /// Indicates to the behaviour that we have discovered a new external address for us.
+    NewExternalAddr { addr: &'a Multiaddr },
+    /// Indicates to the behaviour that an external address was removed.
+    ExpiredExternalAddr { addr: &'a Multiaddr },
+}
+
+impl<'a, Handler: IntoConnectionHandler> InEvent<'a, Handler> {
+    fn map_handler<NewHandler>(
+        self,
+        map_into_handler: impl FnOnce(Handler) -> NewHandler,
+        map_handler: impl FnOnce(
+            <Handler as IntoConnectionHandler>::Handler,
+        ) -> <NewHandler as IntoConnectionHandler>::Handler,
+        map_event: impl FnOnce(
+            <<Handler as IntoConnectionHandler>::Handler as ConnectionHandler>::OutEvent,
+        ) ->
+                <<NewHandler as IntoConnectionHandler>::Handler as ConnectionHandler>::OutEvent,
+    ) -> InEvent<'a, NewHandler>
+    where
+        NewHandler: IntoConnectionHandler,
+    {
+        self.try_map_handler(
+            |h| Some(map_into_handler(h)),
+            |h| Some(map_handler(h)),
+            |e| Some(map_event(e)),
+        )
+        .expect("To return Some as all closures return Some.")
+    }
+
+    fn try_map_handler<NewHandler>(
```

Comment:
Nit: IMO `try_` implies failure, meaning this should return a `Result` instead.

The standard library also follows this: https://doc.rust-lang.org/std/?search=try_

Reply:
If we want it to return an `Option`, perhaps `maybe` might be a better prefix?

Reply:
:+1: done via 46f20fab.